src/components/tl/nccl/tl_nccl_coll.c

```diff
@@ -0,0 +1,187 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_nccl_coll.h"
+#include "utils/ucc_math.h"
+
+#define ncclOpUnsupported (ncclNumOps + 1)
+#define ncclDataTypeUnsupported (ncclNumTypes + 1)
+
+ncclDataType_t ucc_to_nccl_dtype[] = {
+    [UCC_DT_INT8]        = ncclInt8,
+    [UCC_DT_INT16]       = ncclDataTypeUnsupported,
+    [UCC_DT_INT32]       = ncclInt32,
+    [UCC_DT_INT64]       = ncclInt64,
+    [UCC_DT_INT128]      = ncclDataTypeUnsupported,
+    [UCC_DT_UINT8]       = ncclUint8,
+    [UCC_DT_UINT16]      = ncclDataTypeUnsupported,
+    [UCC_DT_UINT32]      = ncclUint32,
+    [UCC_DT_UINT64]      = ncclUint64,
+    [UCC_DT_UINT128]     = ncclDataTypeUnsupported,
+    [UCC_DT_FLOAT16]     = ncclFloat16,
+    [UCC_DT_FLOAT32]     = ncclFloat32,
+    [UCC_DT_FLOAT64]     = ncclFloat64,
+    [UCC_DT_USERDEFINED] = ncclDataTypeUnsupported,
+    [UCC_DT_OPAQUE]      = ncclDataTypeUnsupported,
+};
+
+ncclRedOp_t ucc_to_nccl_reduce_op[] = {
+    [UCC_OP_USERDEFINED] = ncclOpUnsupported,
+    [UCC_OP_SUM]         = ncclSum,
+    [UCC_OP_PROD]        = ncclProd,
+    [UCC_OP_MAX]         = ncclMax,
+    [UCC_OP_MIN]         = ncclMin,
+    [UCC_OP_LAND]        = ncclOpUnsupported,
+    [UCC_OP_LOR]         = ncclOpUnsupported,
+    [UCC_OP_LXOR]        = ncclOpUnsupported,
+    [UCC_OP_BAND]        = ncclOpUnsupported,
+    [UCC_OP_BOR]         = ncclOpUnsupported,
+    [UCC_OP_BXOR]        = ncclOpUnsupported,
+    [UCC_OP_MAXLOC]      = ncclOpUnsupported,
+    [UCC_OP_MINLOC]      = ncclOpUnsupported,
+};
+
+ucc_status_t ucc_nccl_collective_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_nccl_task_t *task = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+    cudaError_t cuda_st;
+
+    cuda_st = cudaEventQuery(task->completed);
+    switch (cuda_st) {
+    case cudaSuccess:
+        coll_task->super.status = UCC_OK;
+        return UCC_OK;
+    case cudaErrorNotReady:
+        return UCC_INPROGRESS;
+    default:
+        return UCC_ERR_NO_MESSAGE;
+    }
+}
+
+ucc_status_t ucc_nccl_alltoall_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_nccl_task_t *task   = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+    ucc_tl_nccl_team_t *team   = task->team;
+    cudaStream_t        stream = team->stream;
+    ucc_rank_t          gsize  = team->size;
+    ucc_status_t        status = UCC_OK;
+    ptrdiff_t           sbuf   = (ptrdiff_t)task->args.src.info.buffer;
+    ptrdiff_t           rbuf   = (ptrdiff_t)task->args.dst.info.buffer;
+    size_t data_size;
+    ucc_rank_t peer;
+
+    task->super.super.status = UCC_INPROGRESS;
+    data_size = (size_t)task->args.src.info.count *
+                ucc_dt_size(task->args.src.info.datatype);
+    NCCLCHECK_GOTO(ncclGroupStart(), exit_coll, status, UCC_TL_TEAM_LIB(team));
+    for (peer = 0; peer < gsize; peer++) {
+        NCCLCHECK_GOTO(ncclSend((void *)(sbuf + peer * data_size), data_size,
```

Comment:
shoudl we check 0 byte NCCL send/recv (https://github.com/pytorch/pytorch/blob/master/torch/csrc/cuda/nccl.cpp#L690)